packages/compass-data-modeling/src/store/relationships.ts

```diff
@@ -0,0 +1,215 @@
+): string[][] {
+  const properties: string[][] = [];
+  traverseMongoDBJSONSchema(schema, (s, path) => {
+    if (s.bsonType === schemaToMatch.bsonType && isEqual(s, schemaToMatch)) {
+      properties.push(path);
+    }
+  });
+  return properties;
+}
+
+/**
+ * @internal exported only for testing purposes
+ */
+export function getValuesFromPath(doc: Document, path: string[]): Document[] {
+  const [currentPath, ...restPath] = path;
+  // We're at the end of the path, return current doc
+  if (!currentPath) {
+    return [doc];
+  }
+  // Path doesn't exist in this document
+  if (!(currentPath in doc)) {
+    return [];
+  }
+  const slice = doc[currentPath];
+  // For arrays, recursively pick up all the values for provided path
+  if (Array.isArray(slice)) {
+    return slice.flatMap((item) => {
+      return getValuesFromPath(item, restPath);
+    });
+  }
+  // Otherwise just continue moving forward through the path
+  return getValuesFromPath(slice, restPath);
+}
+
+/**
+ * A function that is given a starting collection and a list of other
+ * collections in the database returns a list of identified relationships in the
+ * database using the following algorighm:
+ *
+ * For a collection (assumed foreign)
+ *   - If collection doesn’t have an index on _id field, return
+ *   - For each collection (assumed local)
+ *       > If collection name equals the foreign collection name, continue to
+ *         the next collection
+ *       > For every field in local collection
+ *           + If field type matches foreign collection _id type
+ *               * Pick sample values for the field from provided samples
+ *               * Run a count against foreign collection querying by sample
+ *                 values for _id field
+ *               * If the returned count equals the amount of sample values,
+ *                 return relationship
+ *
+ * @param foreignNamespace collection that is assumed "foreign" in the
+ * relationship
+ * @param foreignSchema schema of the "foreign" collection
+ * @param _sampleDocs
+ * @param collections list of all collections that will be checked for matching
+ * relationships
+ * @param dataService dataService instance
+ * @param abortSignal signal to cancel the inferring process
+ * @param onError callback that will be called if inference fails with an error
+ * @returns a list of confirmed relationships
+ */
+export async function inferForeignToLocalRelationshipsForCollection(
+  foreignNamespace: string,
+  foreignSchema: MongoDBJSONSchema,
+  _sampleDocs: Document[],
+  collections: { ns: string; schema: MongoDBJSONSchema; sample: Document[] }[],
+  dataService: DataService,
+  abortSignal?: AbortSignal,
+  onError?: (err: any) => void
+): Promise<Relationship['relationship'][]> {
+  const idSchema = foreignSchema.properties?._id;
+  if (!idSchema) {
+    return [];
+  }
+  const indexes = await dataService
+    .indexes(foreignNamespace, { full: false })
+    .catch(() => {
```

Comment:
Just making sure – we're not logging anything because we know that the `DataService` method already will, correct?